### PR TITLE
Stop and flush on host shutdown

### PR DESCRIPTION
**Pull request by justinmchase**
### The issue or feature being addressed

#659

### Details on the issue fix or feature implementation

I have a CronJob which may start and stop very quickly. Some metrics are generated but as a result of how short lived the application is, the metrics are never sent. After debugging the issue I have concluded that the reason for this is because the reporter is running as a background worker which is essentially cancelled before it can even start. Furthermore due to the way the hosted application works, the TaskFactory never completes once the application is shutting down.

This change uses the `IHostApplicationLifetime` service to detect the `ApplicationStopping` signal and flushes the remaining metrics one last time before the application is fully stopped.

### Confirm the following

- [x] I have ensured that I have merged the latest changes from the dev branch
- [x] I have successfully run a [local build](https://github.com/AppMetrics/AppMetrics#how-to-buil)
- [ ] I have included unit tests for the issue/feature
- [x] I have included the github issue number in my commits

**_Notes:_**

* The build was broken already on a vanilla checkout but the project I edited was building and continues to build fine.
* I did not see existing unit tests for this function or area of code, I'm not sure how to proceed with a unit test.

### CODE REVIEW

File: src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs [hunk 1]
```diff
@@ -60,6 +63,9 @@ public class MetricsReporterBackgroundService : BackgroundService
                         NextRunTime = referenceTime
                     });
             }
+
+            
+            lifetime.ApplicationStopping.Register(() => this.OnApplicationStopping().Wait());
```
Comment (Aaronontheweb) [inline src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:68]: LGTM

File: src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs [hunk 2]
```diff
@@ -97,49 +103,50 @@ private async Task ExecuteOnceAsync(CancellationToken cancellationToken)
 
                 flushTask.Increment();
 
-                await taskFactory.StartNew(
```
Comment (Aaronontheweb) [inline src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: Why was this necessary before? cc @alhardy

**Review by Aaronontheweb — APPROVED** [review]
LGTM but wasn't sure about why we used a custom `Task` context to execute in before - might be something that I don't know about happening behind the scenes there.

File: src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs (hunk 2 above)
Comment (alhardy) [reply src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: The design of the BackgroundService was taken from the example [here](https://blog.maartenballiauw.be/post/2017/08/01/building-a-scheduled-cache-updater-in-aspnet-core-2.html)

"So TaskFactory.StartNew()… Why not simply await them here, you ask? Well, what if you schedule a task to run every minute but that task never returns (or returns after a couple of minutes)? Our scheduler would be useless. So instead we’re spawning tasks outside of our scheduler so at least it can keep its promises"

Comment (Aaronontheweb) [reply src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: So we need to add that back in this PR then?

Comment (alhardy) [reply src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: What was the issue in spawning a new task and using the application lifetime token as the cancellation token?
Just thinking if there was some delay in flushing metrics between intervals some metrics would be lost right? I.e. serialized metrics, including the timestamp, could be pending on each tasks if there is for whatever reason delays/throttling on the tsdb.

Comment (justinmchase) [reply src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: Basically when the host is shutting down and I call the function with `.Wait()` it will then queue this task but the apphost is waiting on all existing Tasks but this new one will basically hit the first await and then never resume. The apphost exits and this just disappears, not even cancelled.

I could add it back and then add a parameter which essentially will optionally do it but it just can't work to start a new task while shutting down.

Comment (alhardy) [reply src/Extensions/src/App.Metrics.Extensions.Hosting/MetricsReporterBackgroundService.cs:100]: OK, let's leave the PR as is.

**Review by alhardy — APPROVED** [review]
